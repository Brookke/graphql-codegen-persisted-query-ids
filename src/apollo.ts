export function usePregeneratedHashed(hashes: {
    [operationsName: string]: string | undefined;
}) {
    return (doc: import("graphql").DocumentNode) => {
        const operationDefinition = doc.definitions.find(
            def => def.kind === "OperationDefinition",
        );

        if (
            !operationDefinition ||
            operationDefinition.kind !== "OperationDefinition"
        ) {
            console.error("Cannot find OperationDefinition from", doc);
            throw new Error("Operation missing from graphql query");
        }

        if (!operationDefinition.name) {
            throw new Error("name missing from operation definition");
        }

        const hash = hashes[operationDefinition.name.value];

        if (!hash) {
            throw new Error(
                "Cannot find pregerated has for " +
                    operationDefinition.name.value,
            );
        }

        return hash;
    };
}
