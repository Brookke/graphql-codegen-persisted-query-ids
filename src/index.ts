import crypto from "crypto";
import {
    print,
    DocumentNode,
    OperationDefinitionNode,
    visit,
    FieldNode,
    FragmentDefinitionNode,
    DefinitionNode,
} from "graphql";
import { PluginFunction } from "graphql-codegen-core";

type Definition = FragmentDefinitionNode | OperationDefinitionNode;

function createHash(s: string) {
    return crypto
        .createHash("sha256")
        .update(s, "utf8")
        .digest()
        .toString("hex");
}

function printDefinitions(definitions: (Definition | DocumentNode)[]) {
    return definitions.map(print).join("\n");
}

const TYPENAME_FIELD: FieldNode = {
    kind: "Field",
    name: {
        kind: "Name",
        value: "__typename",
    },
};

// From apollo-client https://github.com/apollographql/apollo-client/blob/3a9dfe268979618180823eef93e96ab87468449c/packages/apollo-utilities/src/transform.ts
function addTypenameToDocument(doc: DocumentNode): DocumentNode {
    return visit(doc, {
        SelectionSet: {
            enter(node, _key, parent) {
                // Don't add __typename to OperationDefinitions.
                if (
                    parent &&
                    (parent as OperationDefinitionNode).kind ===
                        "OperationDefinition"
                ) {
                    return;
                }

                // No changes if no selections.
                const { selections } = node;
                if (!selections) {
                    return;
                }

                // If selections already have a __typename, or are part of an
                // introspection query, do nothing.
                const skip = selections.some(selection => {
                    return (
                        selection.kind === "Field" &&
                        ((selection as FieldNode).name.value === "__typename" ||
                            (selection as FieldNode).name.value.lastIndexOf(
                                "__",
                                0,
                            ) === 0)
                    );
                });
                if (skip) {
                    return;
                }

                // Create and return a new SelectionSet with a __typename Field.
                return {
                    ...node,
                    selections: [...selections, TYPENAME_FIELD],
                };
            },
        },
    });
}

export interface PluginConfig {
    output: "server" | "client" | undefined;
}

export function findFragmentSpreadsNames(operation: OperationDefinitionNode) {
    const fragmentNames = new Set<string>();

    visit(operation, {
        FragmentSpread: {
            enter(node) {
                fragmentNames.add(node.name.value);
            },
        },
    });

    return fragmentNames;
}

export function findFragments(docs: DocumentNode[]) {
    const fragments: FragmentDefinitionNode[] = [];

    for (const doc of docs) {
        visit(doc, {
            FragmentDefinition: {
                enter(node) {
                    fragments.push(node);
                },
            },
        });
    }

    return fragments;
}

export function generateQueryIds(docs: DocumentNode[], config: PluginConfig) {
    docs = docs.map(addTypenameToDocument);

    const out: { [key: string]: string } = {};

    const fragments = findFragments(docs);

    for (const doc of docs) {
        visit(doc, {
            OperationDefinition: {
                enter(def) {
                    if (!def.name) {
                        throw new Error("OperationDefinition missing name");
                    }

                    const usedFragmentNames = findFragmentSpreadsNames(def);

                    const usedFragments = fragments.filter(frag =>
                        usedFragmentNames.has(frag.name.value),
                    );

                    const query = printDefinitions([...usedFragments, def]);
                    const hash = createHash(query);

                    if (config.output === "client") {
                        out[def.name.value] = hash;
                    } else {
                        out[hash] = query;
                    }
                },
            },
        });
    }

    return out;
}

export const plugin: PluginFunction<PluginConfig> = (
    _schema,
    documents,
    config,
) => {
    if (config.output !== "client" && config.output !== "server") {
        throw new Error(
            "graphql-codegen-persisted-query-id must configure output to 'server' or 'client'",
        );
    }

    const out = generateQueryIds(documents.map(doc => doc.content), config);

    return JSON.stringify(out, null, "   ");
};
