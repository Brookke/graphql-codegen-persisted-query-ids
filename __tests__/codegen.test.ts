import { parse } from "graphql";
import { generateQueryIds, findUsedFragments, findFragments } from "../src";

// Nooop gql fn for prettier
function gql(...things: TemplateStringsArray[]) {
    return things.join("");
}

describe("can generate query for simple doc", () => {
    const doc = parse(gql`
        query Foo {
            bar
        }
    `);

    test("client", () => {
        const client = generateQueryIds([doc], {
            output: "client",
        });

        expect(client["Foo"]).toBeTruthy();
    });

    test("query ids match from client to server", () => {
        const client = generateQueryIds([doc], {
            output: "client",
        });

        const server = generateQueryIds([doc], {
            output: "server",
        });

        const query = server[client["Foo"]];
        expect(query).toBeTruthy();
        expect(query).toMatchSnapshot();
    });
});

describe("fragments", () => {
    const doc1 = parse(gql`
        fragment myFragment on Ding {
            name
        }

        query Foo {
            bar
            ...myFragment
        }
    `);

    const doc2 = parse(gql`
        query Foo2 {
            bar
        }
    `);

    const multiQueryDoc = parse(gql`
        fragment myFragment on Ding {
            name
        }

        query Foo {
            bar
            ...myFragment
        }

        query Foo2 {
            bar
        }
    `);

    test("is added to queries", () => {
        const server = generateQueryIds([doc1, doc2], {
            output: "server",
        });

        const client = generateQueryIds([doc1, doc2], {
            output: "client",
        });

        expect(server[client["Foo"]]).toContain("fragment myFragment");
    });

    test("multiple docs not using fragments", () => {
        const server = generateQueryIds([doc1, doc2], {
            output: "server",
        });

        const client = generateQueryIds([doc1, doc2], {
            output: "client",
        });

        expect(server[client["Foo2"]]).not.toContain("fragment myFragment");
    });

    test("multi query doc", () => {
        const server = generateQueryIds([multiQueryDoc], {
            output: "server",
        });

        const client = generateQueryIds([multiQueryDoc], {
            output: "client",
        });

        expect(server[client["Foo"]]).toContain("fragment myFragment");
        expect(server[client["Foo2"]]).not.toContain("fragment myFragment");
    });

    test("can use fragment before it's definition", () => {
        const doc = parse(gql`
            query Foo {
                bar
                ...myFragment
            }

            fragment myFragment on Ding {
                name
            }
        `);

        const server = generateQueryIds([doc], {
            output: "server",
        });

        const client = generateQueryIds([doc], {
            output: "client",
        });

        const query = server[client["Foo"]];
        expect(query).toBeTruthy();
    });

    test("fragments in fragments work", () => {
        const fragInFrag = parse(gql`
            fragment nestedFrag on Ding {
                fromNested
            }

            fragment myFragment on Ding {
                name
                ...nestedFrag
            }

            query Foo {
                bar
                ...myFragment
            }
        `);

        const server = generateQueryIds([fragInFrag], {
            output: "server",
        });

        const client = generateQueryIds([fragInFrag], {
            output: "client",
        });

        const query = server[client["Foo"]];
        expect(query).toBeTruthy();
        expect(query).toContain("fragment nestedFrag");
        expect(query).toContain("fragment myFragment");
    });
});

describe("mutation", () => {
    const doc = parse(gql`
        mutation AddTodo($title: String!) {
            createTodo(
                input: {
                    title: $title
                    clientMutationId: "lala"
                    completed: false
                    status: PUBLISH
                }
            ) {
                clientMutationId
                todo {
                    id
                    title
                    completed
                }
            }
        }
    `);

    test("query ids match from client to server", () => {
        const client = generateQueryIds([doc], {
            output: "client",
        });

        const server = generateQueryIds([doc], {
            output: "server",
        });

        const query = server[client["AddTodo"]];
        expect(query).toBeTruthy();
    });
});

test("can find nested fragment user", () => {
    const doc = parse(gql`
        fragment TodoParts on Todo {
            title
        }

        query DualTodoList($cursorTodos: String!, $cursorDones: String!) {
            todos(first: 3, after: $cursorTodos, where: { completed: false }) {
                ...TodoParts
            }
            dones: todos(
                first: 3
                after: $cursorDones
                where: { completed: true }
            ) {
                ...TodoParts
            }
        }
    `);

    const operation = doc.definitions.find(
        def => def.kind === "OperationDefinition",
    );

    if (!operation || operation.kind !== "OperationDefinition") {
        throw new Error("cannot find operation");
    }

    const knownFragments = findFragments([doc]);

    const fragmentNames = Array.from(
        findUsedFragments(operation, knownFragments).values(),
    ).map(frag => frag.name.value);

    expect(fragmentNames).toEqual(["TodoParts"]);
});
