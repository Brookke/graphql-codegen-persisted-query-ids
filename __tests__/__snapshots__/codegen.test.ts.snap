// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutations query ids match from client to server 1`] = `
"mutation AddTodo($title: String!) {
  createTodo(input: {title: $title, clientMutationId: \\"lala\\", completed: false, status: PUBLISH}) {
    clientMutationId
    todo {
      id
      title
      completed
    }
  }
}"
`;

exports[`can generate query for simple doc query ids match from client to server 1`] = `
"query Foo {
  bar
}"
`;
